### PR TITLE
Taxonomy: freeform interaction and base infrastructure

### LINKED ISSUE #8838: Taxonomy: Add tags
Sets up the initial tag section of the new section of the document panel with the ability to add a tag. 

See  #8830 for API details. 

Note that some users will have custom freeform taxonomies in their site. These will have the same structure as the baked-in tag structure, so try to keep that in mind here so we can support custom taxonomies here: #8849 

## Acceptance criteria 
- Can type a new tag and saves when comma or enter is pressed, updating the text into a tag. 
- Add tests

### LINKED ISSUE #8839: Taxonomy: Remove tags
Should be done after #8838  
Adds functionality to remove tags from a story

See #8830 for API details.

## Acceptance criteria 
- Clicking on the 'x' of a tag should delete it from the story. 
- I should be able to tab through the list of tags and get focus put onto each button, using the keyboard should also delete a tag from a story. 
- Test added

### PR DESCRIPTION
**Pull request by maxyinger**
## Context
Part of initial taxonomy FE infrastructure and WP API interactions.

## Summary
Adds: 
- Base Taxonomy provider setup
- Hydrating initial taxonomy from loaded story
- Updating taxonomy in story state
- Saving Taxonomy when saving story
- Creating feeform taxonomy terms
- Tying together freeform taxonomy input with taxonomy BE interactions

## Relevant Technical Choices

**Shape of State**
Basic idea here is that we store all taxonomies in a nested dictionary/cache. Shape should follow:
```
TermCache {
  [Taxonomy['rest_base']]: {
    [Term['slug]]: Term
  }
}
```

We keep track of selected/post associated terms in the following shape:
```
SelectedSlugs {
  [Taxonomy['rest_base]]: Array<Term['slug']>
}
```

**Initial Hydration**
Terms are pulled off the story's `_embed['wp:term']` property on initial load and used to populate the initial cache as well as the initial selected slugs.

**User Interaction**
Terms are fetched from the BE and populate the cache from either:
  - Debounced callback to freeform input requesting `/<taxonomy>?search=<current_input_val>`
  - Post to create tag on submission of freeform input

This allows us to update selected slugs in real time without waiting for the response from the BE, and we sync up the story's selected term IDs every time either the Selected Slugs or Term Cache update.

## To-do
NA

Gonna increase test coverage here. karma and provider unit:
https://app.zenhub.com/workspaces/web-stories-5e94d3aced449034e1e9b226/issues/google/web-stories-wp/8842

## User-facing changes
NA

## Testing Instructions
- Upload this plugin and activate it to register a couple taxonomies onto the story post type:
[web-stories-taxonomy-example.php.zip](https://github.com/google/web-stories-wp/files/7164177/web-stories-taxonomy-example.php.zip)

- Make sure the taxonomy flag is enabled.
- Go to editor & open up the document panel
- scroll down to tags & categories panel
- add some taxonomies
- save story
- refresh page
- see that taxonomies are persisted

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8838 
Fixes #8839

### CODE REVIEW

File: packages/story-editor/src/app/api/apiProvider.js [hunk 1]
```diff
@@ -501,16 +503,35 @@ function APIProvider({ children }) {
...
     [customPageTemplates]
   );
 
+  const getTaxonomies = useCallback(() => {
+    return apiFetch({
+      path: addQueryArgs('/wp/v2/taxonomies/', { type: 'web-story' }),
+    });
+  }, []);
+
+  const getTaxonomicTerm = useCallback((taxonomy, args = {}) => {
```
Comment (swissspidy) [inline packages/story-editor/src/app/api/apiProvider.js]:
```suggestion
  const getTaxonomyTerm = useCallback((taxonomy, args = {}) => {
```

File: packages/story-editor/src/app/api/apiProvider.js [hunk 2]
```diff
@@ -501,16 +503,35 @@ function APIProvider({ children }) {
...
+    return apiFetch({
+      path: addQueryArgs(`/wp/v2/${taxonomy}`, args),
+    });
+  }, []);
+
+  const createTaxonomicTerm = useCallback((taxonomy, name) => {
```
Comment (swissspidy) [inline packages/story-editor/src/app/api/apiProvider.js]:
```suggestion
  const createTaxonomyTerm = useCallback((taxonomy, name) => {
```

File: packages/story-editor/src/constants/index.js [hunk 3]
```diff
@@ -96,3 +96,5 @@ export const OUTLINK_THEME = {
   LIGHT: 'light',
   CUSTOM: 'custom',
 };
+
+export const MAXIMUM_ENTRIES_PER_PAGE = 100;
```
Comment (swissspidy) [inline packages/story-editor/src/constants/index.js]: This is not used anywhere? Also a bit of an ambiguous name.

File: packages/story-editor/src/app/api/apiProvider.js [hunk 4]
```diff
@@ -501,16 +503,35 @@ function APIProvider({ children }) {
       // `?_method=DELETE` is an alternative solution to override the request method.
       // See https://developer.wordpress.org/rest-api/using-the-rest-api/global-parameters/#_method-or-x-http-method-override-header
       return apiFetch({
-        path: addQueryArgs(`${customPageTemplates}${id}/`, {
-          _method: 'DELETE',
-        }),
-        data: { force: true },
+        path: addQueryArgs(`${customPageTemplates}${id}/`, {}),
+        data: {},
```
Comment (swissspidy) [inline packages/story-editor/src/app/api/apiProvider.js]: Why was this changed?

File: packages/story-editor/src/app/api/apiProvider.js [hunk 5]
```diff
@@ -53,9 +53,11 @@ const STORY_FIELDS = [
   'permalink_template',
   'style_presets',
   'password',
+  'story-tags',
 ].join(',');
 
-const STORY_EMBED = 'wp:featuredmedia,wp:lockuser,author,wp:publisherlogo';
+const STORY_EMBED =
+  'wp:featuredmedia,wp:lockuser,author,wp:publisherlogo,wp:term';
```
Comment (swissspidy) [inline packages/story-editor/src/app/api/apiProvider.js]: I think we also need to update that here:

https://github.com/google/web-stories-wp/blob/9976e33b321b60aea81b219599b97f53fc83854c/includes/templates/admin/edit-story.php#L45

File: packages/story-editor/src/app/api/apiProvider.js [hunk 6]
```diff
@@ -501,16 +503,35 @@ function APIProvider({ children }) {
...
 
+  const getTaxonomies = useCallback(() => {
+    return apiFetch({
+      path: addQueryArgs('/wp/v2/taxonomies/', { type: 'web-story' }),
+    });
+  }, []);
+
+  const getTaxonomicTerm = useCallback((taxonomy, args = {}) => {
+    return apiFetch({
+      path: addQueryArgs(`/wp/v2/${taxonomy}`, args),
```
Comment (swissspidy) [inline packages/story-editor/src/app/api/apiProvider.js]:
```suggestion
      path: addQueryArgs(`/wp/v2/${taxonomy}/`, args),
```

File: packages/story-editor/src/app/api/apiProvider.js [hunk 7]
```diff
@@ -501,16 +503,35 @@ function APIProvider({ children }) {
...
+    });
+  }, []);
+
+  const createTaxonomicTerm = useCallback((taxonomy, name) => {
+    return apiFetch({
+      path: addQueryArgs(`/wp/v2/${taxonomy}`, {
```
Comment (swissspidy) [inline packages/story-editor/src/app/api/apiProvider.js]:
```suggestion
      path: addQueryArgs(`/wp/v2/${taxonomy}/`, {
```

File: packages/story-editor/src/app/api/apiProvider.js (hunk 4 above)
Comment (spacedmonkey) [reply packages/story-editor/src/app/api/apiProvider.js]: +1 to me, this doesn't make sense.

File: packages/story-editor/src/app/api/apiProvider.js [hunk 8]
```diff
@@ -501,16 +503,35 @@ function APIProvider({ children }) {
...
+        data: {},
         method: 'POST',
       });
     },
     [customPageTemplates]
   );
 
+  const getTaxonomies = useCallback(() => {
+    return apiFetch({
+      path: addQueryArgs('/wp/v2/taxonomies/', { type: 'web-story' }),
```
Comment (spacedmonkey) [inline packages/story-editor/src/app/api/apiProvider.js]: All other endpoints, using the web-stories/v1/ namespace, which gives us more control. We should do this here as well @swissspidy ?

Comment (swissspidy) [reply packages/story-editor/src/app/api/apiProvider.js]: Perhaps, let's discuss in a separate ticket+PR

File: packages/story-editor/src/app/story/effects/useLoadStory.js [hunk 9]
```diff
@@ -100,6 +103,10 @@ function useLoadStory({ storyId, shouldLoad, restore, isDemo }) {
           url: embedded?.['wp:publisherlogo']?.[0]?.source_url || '',
         };
 
+        if ('wp:term' in embedded) {
```
Comment (spacedmonkey) [inline packages/story-editor/src/app/story/effects/useLoadStory.js]: There will need read values in the link to check if the current use assign / create terms. See
Screenshot 2021-09-13 at 15 05 03

If user does not have assign / create action link here, the assign / create will fail. So these options will need to hidden.

Comment (spacedmonkey) [reply packages/story-editor/src/app/story/effects/useLoadStory.js]: More context https://github.com/google/web-stories-wp/pull/8591 

and https://github.com/google/web-stories-wp/blob/f62effbfc605a3e2a4ddbc8cfd9fa707db28c57f/packages/story-editor/src/app/story/effects/useLoadStory.js#L65-L74

So we will need entries in this object like this `hasAssignTagAction` and `hasCreateTagAction`. But they will need to dynamic of course.

File: packages/story-editor/src/app/api/apiProvider.js (hunk 4 above)
Comment (maxyinger) [reply packages/story-editor/src/app/api/apiProvider.js]: yea I don't know why this was changed either. will revert

File: packages/story-editor/src/app/api/apiProvider.js (hunk 8 above)
Comment (maxyinger) [reply packages/story-editor/src/app/api/apiProvider.js]: yea open to whatever. I'm not sure what the difference is between v1 & v2 here

Comment (spacedmonkey) [reply packages/story-editor/src/app/api/apiProvider.js]: We should preload this request on the page.

Comment (spacedmonkey) [reply packages/story-editor/src/app/api/apiProvider.js]: https://github.com/google/web-stories-wp/blob/f62effbfc605a3e2a4ddbc8cfd9fa707db28c57f/includes/templates/admin/edit-story.php#L39-L111

File: packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js [hunk 10]
```diff
@@ -18,18 +18,124 @@
...
       {...props}
-    />
+    >
+      {taxonomyHandlerTuples.map(([taxonomy, handlers]) =>
+        // TODO support all taxonomies and differentiate
+        // input component based on `taxonomy.hierarchical`
+        taxonomy.slug === 'story-tag' ? (
+          <div key={taxonomy.slug}>
+            <Tags.Label htmlFor={`${taxonomy.slug}-input`}>
+              {__('Add New Term', 'web-stories')}
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js]: Does this need to respect the labels register the taxonomy labels. See https://developer.wordpress.org/reference/functions/get_taxonomy_labels/#return

Comment (spacedmonkey) [reply packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js]: These labels could be added to the taxononmies endpoint if that is at all helpful.

File: packages/story-editor/src/constants/index.js (hunk 3 above)
Comment (maxyinger) [reply packages/story-editor/src/constants/index.js]: Ah yeaup. was part of an earlier approach. thanks for flagging

File: packages/story-editor/src/app/story/effects/useLoadStory.js (hunk 9 above)
Comment (maxyinger) [reply packages/story-editor/src/app/story/effects/useLoadStory.js]: know we already talked on slack, but posting new ticket for this here for book keeping. Thanks again for flagging this 🙌 

#9011

File: packages/story-editor/src/app/api/apiProvider.js [hunk 11]
```diff
@@ -53,9 +53,11 @@ const STORY_FIELDS = [
   'permalink_template',
   'style_presets',
   'password',
+  'story-tags',
```
Comment (maxyinger) [inline packages/story-editor/src/app/api/apiProvider.js]: @swissspidy @spacedmonkey I think we're fine without this since all associated terms should be in the `_embed` . Is it fine just to use the `_embed['wp:term']` as the reference for all associated terms?

Screen Shot 2021-09-13 at 12 01 31 PM

File: packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js (hunk 10 above)
Comment (maxyinger) [reply packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js]: Oh yes. I was looking for something like this and just had this in here temporarily. Thanks!

Comment (maxyinger) [reply packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js]: yea adding them to the endpoint would be awesome. Will be easy to update here if it's already in the taxonomies response

File: packages/story-editor/src/app/api/apiProvider.js (hunk 8 above)
Comment (maxyinger) [reply packages/story-editor/src/app/api/apiProvider.js]: would I need to update anything on the client? or does `apiFetch` do all the work to grab a preloaded request if it exists?

File: packages/story-editor/src/app/api/apiProvider.js (hunk 11 above)
Comment (swissspidy) [reply packages/story-editor/src/app/api/apiProvider.js]: Yes

File: packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js (hunk 10 above)
Comment (swissspidy) [reply packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js]: The labels information is already available from the taxonomies endpoint. No need to reinvent the wheel. You just need to pass `&context=edit` in the `getTaxonomies` API request to get all the taxonomies including the labels.

With edit context, the user needs to have the `assign_terms` capability for a taxonomy to see it & its labels. This relates a bit to #9011

File: packages/story-editor/src/app/api/apiProvider.js (hunk 8 above)
Comment (swissspidy) [reply packages/story-editor/src/app/api/apiProvider.js]: No, you don't. Yes, it does.

But preloading is a different topic altogether.

@spacedmonkey Could you create separate tickets for these? Thanks

Comment (spacedmonkey) [reply packages/story-editor/src/app/api/apiProvider.js]: I wrote a PR as it was so quick. https://github.com/google/web-stories-wp/pull/9032

**Comment by spacedmonkey** [comment]
This will need to be rebased, now that https://github.com/google/web-stories-wp/pull/8966 was merged.

File: packages/story-editor/src/app/story/utils/getStoryPropsToSave.js [hunk 12]
```diff
@@ -35,6 +35,9 @@ function getStoryPropsToSave({ story, pages, metadata }) {
     'autoAdvance',
     'defaultPageDuration',
     'backgroundAudio',
+    // TODO woof. need to make this dynamic based on
+    // taxonomies returned from WP at some point.
```
Comment (BrittanyIRL) [inline packages/story-editor/src/app/story/utils/getStoryPropsToSave.js]: Is this associated with a ticket?

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+  const [taxonomies, setTaxonomies] = useState([]);
+  const [selectedSlugs, setSelectedSlugs] = useState({});
+  const [termCache, setTermCache] = useState({});
+  const { updateStory, isStoryLoaded, story } = useStory(
+    ({ state: { pages, story }, actions: { updateStory } }) => ({
+      updateStory,
+      isStoryLoaded: pages.length > 0,
+      story,
+    })
+  );
```
Comment (BrittanyIRL) [inline packages/story-editor/src/app/taxonomy/taxonomyProvider.js]: 👍

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+  // Get all registered `web-story` taxonomies.
+  useEffect(() => {
+    (async function () {
+      try {
+        const result = await getTaxonomies();
+        setTaxonomies(Object.values(result));
+      } catch (e) {
+        // Do we wanna do anything here?
```
Comment (BrittanyIRL) [inline packages/story-editor/src/app/taxonomy/taxonomyProvider.js:70]: Made a new ticket to research this: https://github.com/google/web-stories-wp/issues/9033

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+  // Reference embedded terms in the story and taxonomies
+  // to get the initial selected terms as well as populate
+  // the taxonomy term cache
+  const hasHydratedRunOnce = useRef(false);
+  useEffect(() => {
+    if (taxonomies.length > 0 && isStoryLoaded && !hasHydratedRunOnce.current) {
+      const taxonomiesBySlug = dictonaryOnKey(taxonomies, 'slug');
+      const initialCache = mapObjectKeys(
+        cacheFromEmbeddedTerms(story.embeddedTerms),
+        (slug) => taxonomiesBySlug[slug].rest_base
```
Comment (BrittanyIRL) [inline packages/story-editor/src/app/taxonomy/taxonomyProvider.js]: ??

```suggestion
        (slug) => taxonomiesBySlug[slug]?.rest_base
```

**Comment by maxyinger** [comment]
@swissspidy @spacedmonkey do we want to register the `story-tags` and `story-category` taxonomies directly in the plugin? just wanted to mention cus I forgot about that bit when writing testing instructions

**Review by BrittanyIRL — APPROVED** [review]
Really great comments and I like the /utils set up that you did here - there's so much structuring that needs to happen, i like that you kept it all together and so tight. 
Works as you specify. 

Want a new ticket for predictive suggestions of existing tags? I think I missed that functionality when I initially sketched out issues.
Screen Shot 2021-09-14 at 11 19 24 AM

Once all the tests are passing I'm good with this - I know it's blocking further work for you and @samwhale

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+} from './utils';
+
+function TaxonomyProvider(props) {
+  const [taxonomies, setTaxonomies] = useState([]);
+  const [selectedSlugs, setSelectedSlugs] = useState({});
+  const [termCache, setTermCache] = useState({});
+  const { updateStory, isStoryLoaded, story } = useStory(
+    ({ state: { pages, story }, actions: { updateStory } }) => ({
+      updateStory,
+      isStoryLoaded: pages.length > 0,
```
Comment (samwhale) [inline packages/story-editor/src/app/taxonomy/taxonomyProvider.js:49]: This is cool 👍

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js (hunk 14 above)
Comment (samwhale) [inline packages/story-editor/src/app/taxonomy/taxonomyProvider.js:70]: 💭 If this fails maybe we hide the taxonomy UI and log an error?

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+      } catch (e) {
+        // Do we wanna do anything here?
+      }
+    })();
+  }, [getTaxonomies]);
+
+  // Reference embedded terms in the story and taxonomies
+  // to get the initial selected terms as well as populate
+  // the taxonomy term cache
+  const hasHydratedRunOnce = useRef(false);
```
Comment (samwhale) [inline packages/story-editor/src/app/taxonomy/taxonomyProvider.js]: 💭 nit: this feels a little more grammatically correct to me, but I don't think it matters
```suggestion
  const hasHydrationRunOnce = useRef(false);
```

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+  }, [updateStory, selectedSlugs, termCache]);
+
+  const addSearchResultsToCache = useCallback(
+    async (taxonomy, name) => {
+      let response = [];
+      try {
+        response = await getTaxonomyTerm(taxonomy.rest_base, {
+          search: name,
+          // This is the per_page value Gutenberg is using
+          per_page: 20,
```
Comment (samwhale) [inline packages/story-editor/src/app/taxonomy/taxonomyProvider.js:136]: 💭 maybe we can set this as a default in the options but allow a request to be sent with more (100 is the max right?)

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+        });
+      } catch (e) {
+        // Do we wanna do anything here?
+      }
+
+      // Avoid update if we're not actually adding any terms here
+      if (response.length < 1) {
+        return;
+      }
```
Comment (samwhale) [inline packages/story-editor/src/app/taxonomy/taxonomyProvider.js:145]: :smart:

File: packages/story-editor/src/app/taxonomy/utils/index.js [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+ */
+export function mapObjectKeys(obj = {}, op = (v) => v) {
+  return objectFromEntries(
+    Object.entries(obj).map(([key, val]) => [op(key), val])
+  );
+}
+
+/**
+ * Merges two object who have children that can't be
+ * merged by spreading both object into a new object.
```
Comment (samwhale) [inline packages/story-editor/src/app/taxonomy/utils/index.js]:
```suggestion
 * merged by spreading both objects into a new object.
```

File: packages/story-editor/src/app/taxonomy/utils/index.js [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * @param {Object} dictA nested dictionary
+ * @param {Object} dictB nested dictionary
+ * @return {Object} new unified nested dictionary
+ */
+export function mergeNestedDictionaries(dictA = {}, dictB = {}) {
+  return Object.entries(dictA).reduce((merged, [key, val]) => {
+    merged[key] = { ...val, ...merged[key] };
+    return merged;
+  }, dictB);
```
Comment (samwhale) [inline packages/story-editor/src/app/taxonomy/utils/index.js]: 💭 should this reduce begin with a new copy of `dictB` so it doesn't modify the original arg?
```suggestion
  }, { ...dictB });
```

File: packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js [hunk 22]
```diff
@@ -18,18 +18,124 @@
...
+
+  const _termDisplayTransformer = useCallback(
+    (taxonomy) => (tagName) =>
+      termCache[taxonomy]?.[cleanForSlug(tagName)]?.name,
+    [termCache]
+  );
+
+  // We want to prevent curried functions from creating
+  // a new function on every render so we build them with
+  // memoized args here instead of in the render
```
Comment (samwhale) [inline packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js]: 👍

**Review by samwhale — APPROVED** [review]
Nice dude - this is great

**Comment by samwhale** [comment]
> Want a new ticket for predictive suggestions of existing tags? I think I missed that functionality when I initially sketched out issues.
> Screen Shot 2021-09-14 at 11 19 24 AM

I think this work would be good as a separate ticket

**Comment by BrittanyIRL** [comment]
> > Want a new ticket for predictive suggestions of existing tags? I think I missed that functionality when I initially sketched out issues.
> > Screen Shot 2021-09-14 at 11 19 24 AM
> 
> I think this work would be good as a separate ticket

Added here: https://github.com/google/web-stories-wp/issues/9034

File: packages/story-editor/src/app/story/utils/getStoryPropsToSave.js (hunk 12 above)
Comment (maxyinger) [reply packages/story-editor/src/app/story/utils/getStoryPropsToSave.js]: I've been associating it with this ticket to make the panel work for any taxonomy here:
https://app.zenhub.com/workspaces/web-stories-5e94d3aced449034e1e9b226/issues/google/web-stories-wp/8849

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js (hunk 14 above)
Comment (maxyinger) [reply packages/story-editor/src/app/taxonomy/taxonomyProvider.js:70]: Brittany created this ticket here for this behavior:
https://github.com/google/web-stories-wp/issues/9033

File: packages/story-editor/src/app/taxonomy/utils/index.js (hunk 21 above)
Comment (maxyinger) [reply packages/story-editor/src/app/taxonomy/utils/index.js]: ah yeaup. good catch

File: packages/story-editor/src/app/story/utils/getStoryPropsToSave.js [hunk 23]
```diff
@@ -35,6 +35,9 @@ function getStoryPropsToSave({ story, pages, metadata }) {
     'autoAdvance',
     'defaultPageDuration',
     'backgroundAudio',
+    // TODO woof. need to make this dynamic based on
```
Comment (BrittanyIRL) [inline packages/story-editor/src/app/story/utils/getStoryPropsToSave.js]:
```suggestion
    // TODO #8849. Need to make this dynamic based on
```

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js (hunk 18 above)
Comment (maxyinger) [reply packages/story-editor/src/app/taxonomy/taxonomyProvider.js:136]: Yup, 100 is the max, and WP sets the default to 10 if you don't specify.

We can update the function to make the default 20, but feels kinda like arbitrarily moving things around since I'm thinking the value of 20 results per page will be unique behavior to the scope of this function.

Also kinda wanted to minimize making `getTaxonomyTerm` opinionated over the actual query until we see any patterns arise from our use of it.

anyway, happy to update, don't feel strongly either way, but just wanted to shed some light on my rationale.

File: packages/story-editor/src/app/story/utils/getStoryPropsToSave.js (hunk 23 above)
Comment (maxyinger) [reply packages/story-editor/src/app/story/utils/getStoryPropsToSave.js]: ah yea. good call. the ticket id is a much better identifier than `woof`

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js (hunk 18 above)
Comment (samwhale) [reply packages/story-editor/src/app/taxonomy/taxonomyProvider.js:136]: makes sense - let's keep it!

File: packages/story-editor/src/app/story/utils/getStoryPropsToSave.js (hunk 23 above)
Comment (samwhale) [reply packages/story-editor/src/app/story/utils/getStoryPropsToSave.js]: I liked the `woof`

**Comment by samwhale** [comment]
@googlebot I consent.

File: packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js [hunk 24]
```diff
@@ -157,6 +240,31 @@ function TaxonomiesPanel({ ...props }) {
           </LinkButton>
         )}
       </ContentArea>
+
+      <ContentHeading>{__('Tags', 'web-stories')}</ContentHeading>
```
Comment (samwhale) [inline packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js]: Gave it a little content heading :sparkles:

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+  const { getTaxonomyTerm, createTaxonomyTerm, getTaxonomies } = useAPI(
+    ({ actions }) => actions
+  );
+
+  // Get all registered `web-story` taxonomies.
+  useEffect(() => {
+    (async function () {
+      try {
+        const result = await getTaxonomies();
+        setTaxonomies(Object.values(result));
```
Comment (swissspidy) [inline packages/story-editor/src/app/taxonomy/taxonomyProvider.js:68]: I just realized that this causes HTTP 403 responses when no taxonomies are registered (i.e. when you don't have my example plugin installed). That's why currently all e2e tests are failing.

So we need to make this smarter from the beginning. I'll try to whip something up.

Related: #8849

**Comment by swissspidy** [comment]
@littlemilkstudio FYI: With your force push you appear to have completely reverted #9032 again which was merged into this branch... I'll try to restore it.

**Comment by swissspidy** [comment]
@googlebot I consent

File: packages/story-editor/src/app/taxonomy/taxonomyProvider.js (hunk 25 above)
Comment (swissspidy) [reply packages/story-editor/src/app/taxonomy/taxonomyProvider.js:68]: @littlemilkstudio #9037 should fix this

Comment (maxyinger) [reply packages/story-editor/src/app/taxonomy/taxonomyProvider.js:68]: Ahh excellent catch. was gonna trouble shoot that this morning, but thanks for taking the time to figure it out!

**Comment by maxyinger** [comment]
> @littlemilkstudio FYI: With your force push you appear to have completely reverted #9032 again which was merged into this branch... I'll try to restore it.

Ahh glazed over that and thought Johnny had merged the preload PR into main, not this branch. my bad. Thanks for restoring!

File: includes/Admin/Editor.php [hunk 26]
```diff
@@ -324,6 +324,8 @@ public function get_editor_settings(): array {
 					'hotlink'       => '/web-stories/v1/hotlink/',
 					'link'          => '/web-stories/v1/link/',
 					'statusCheck'   => '/web-stories/v1/status-check/',
+					'taxonomies'    => '/wp/v2/taxonomies/',
+					'taxonomyBase'  => '/wp/v2/',
```
Comment (swissspidy) [inline includes/Admin/Editor.php]: Why was this changed back? Should be `wp/v1/taxonomies`

Comment (maxyinger) [reply includes/Admin/Editor.php]: I'm getting 404s for everything else we've tried in here

Comment (maxyinger) [reply includes/Admin/Editor.php]: I'll try v1 again rq to double check

File: packages/story-editor/src/app/api/apiProvider.js [hunk 27]
```diff
@@ -41,10 +41,13 @@ function APIProvider({ children }) {
       currentUser,
       storyLocking,
       pageTemplates: customPageTemplates,
+      taxonomies,
+      taxonomyBase,
```
Comment (swissspidy) [inline packages/story-editor/src/app/api/apiProvider.js]: `taxonomyBase` should not be retrieved from the config like this.

The full REST API URL for each taxonomy can be retrieved from the `_links` section of the `getTaxonomies()` response.

File: includes/Admin/Editor.php (hunk 26 above)
Comment (maxyinger) [reply includes/Admin/Editor.php]: Screen Shot 2021-09-15 at 1 17 05 PM
Screen Shot 2021-09-15 at 1 17 10 PM

Comment (swissspidy) [reply includes/Admin/Editor.php]: @littlemilkstudio You might need to run `composer install` first so the PHP classes are updated.

If that doesn't help, please ping me on Slack.

File: packages/story-editor/src/app/api/apiProvider.js (hunk 27 above)
Comment (maxyinger) [reply packages/story-editor/src/app/api/apiProvider.js]: awesome. Updated it to use that endpoint present in `_links`
